### DISCUSSION TITLE
Hamiltonian or Lagrangian formulation of CFMMs

### DISCUSSION THREAD

**Discussion by poliwop** [galacticcouncil/HydraDX-simulations#4] [Ideas]
It'd be interesting to know if the math behind CFMMs can be structured in a Hamiltonian/Lagrangian framework? (My physics background is far enough in my past I don't know which of these would make more sense...)

There is no useful sense of time in the fundamental CFMM math, but I wonder if replacing this with the numeraire makes sense? If R is the quantity of an asset and Q is the quantity of the numeraire, derivatives dR/dQ already play a fundamental role in the AMM (as spot prices)... and R * dQ/dR also plays a fundamental role (total value of the asset in the system)...

**Comment by RudolfGolubich** [comment]
This is the Forum you wrote about? :-)

The Hamiltonian (or Lagrangian) has in most cases no explicit notion of time, but "generalized" impulses: interdependencies of variables and the corresponding changes of them. In a sense, the notion of invariance as used in AMMs is just one half of a full Hamiltonian: the dynamics is missing. 

Hamiltonian mechanics can (in most cases) be easily transformed into Lagrangian and vice versa: a so called "Legendre transformation". Some things are easier in Hamilton formalism, but difficult in Lagrange, for other it is the opposite.

With best regards,
Rudi

**Reply by poliwop** [reply]
I see, in order to have interesting Hamiltonian mechanics for an AMM we can't just say "treat spot prices like the second set of coordinates", we would actually have to identify a CFMM function which depended explicitly on spot prices rather than have them derived from token quantities.

Or to put it another way, it'd be a much less interesting universe if every particle's velocity was determined entirely by it's position :-)

**Reply by RudolfGolubich** [reply]
Yes, that hits the spot. Luckily, an AMM and its environment are non-deterministic enough to be interesting: I already found some similarities to gauge field theories. 

Maybe one can reduce the formalism to the essentials to allow the price definition to be derived top down, instead of setting it bottom up: by going from configuration space to phase space, a trade-size dependent price function may arise as a kind of coupling of the generalised coordinates to the generalised impulses. With a bit of luck, this could give rise to a less ambiguous value definition, but I suspect, that the price function somehow resembles a gauge dependent observable. The interesting question may be: what are the "good" generalised impulses?
